pkg/projects/resource_project_user_test.go

```diff
@@ -0,0 +1,171 @@
+			key = "{{ .project_key }}"
+			display_name = "{{ .name }}"
+			description = "test description"
+			admin_privileges {
+				manage_members = true
+				manage_resources = true
+				index_resources = true
+			}
+
+			user {
+				name = "{{ .username1 }}"
+				roles = ["{{ .developeRole }}", "{{ .contributorRole }}"]
+			}
+
+			user {
+				name = "{{ .username2 }}"
+				roles = ["{{ .contributorRole }}"]
+			}
+		}
+	`, params)
+
+	noUserConfig := executeTemplate("TestAccProjectUser", `
+		resource "project_project" "{{ .name }}" {
+			key = "{{ .project_key }}"
+			display_name = "{{ .name }}"
+			description = "test description"
+			admin_privileges {
+				manage_members = true
+				manage_resources = true
+				index_resources = true
+			}
+		}
+	`, params)
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() {
+			testAccPreCheck(t)
+			createTestUser(t, projectKey, username1, email1)
+			createTestUser(t, projectKey, username2, email2)
+		},
+		CheckDestroy: verifyDeleted(resourceName, func(id string, request *resty.Request) (*resty.Response, error) {
+			deleteTestUser(t, projectKey, username1)
+			deleteTestUser(t, projectKey, username2)
+			resp, err := verifyProject(id, request)
+
+			return resp, err
+		}),
+		ProviderFactories: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: initialConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", fmt.Sprintf("%s", params["project_key"])),
+					resource.TestCheckResourceAttr(resourceName, "display_name", name),
+					resource.TestCheckResourceAttr(resourceName, "description", "test description"),
+					resource.TestCheckResourceAttr(resourceName, "user.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "user.0.name", username1),
+					resource.TestCheckResourceAttr(resourceName, "user.0.roles.#", "1"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "user.0.roles.*", developeRole),
+				),
+			},
+			{
+				Config: addUserConfig,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "key", fmt.Sprintf("%s", params["project_key"])),
+					resource.TestCheckResourceAttr(resourceName, "display_name", name),
+					resource.TestCheckResourceAttr(resourceName, "description", "test description"),
+					resource.TestCheckResourceAttr(resourceName, "user.#", "2"),
+					resource.TestCheckResourceAttr(resourceName, "user.0.name", username1),
+					resource.TestCheckResourceAttr(resourceName, "user.0.roles.#", "2"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "user.0.roles.*", developeRole),
+					resource.TestCheckTypeSetElemAttr(resourceName, "user.0.roles.*", contributorRole),
+					resource.TestCheckResourceAttr(resourceName, "user.1.name", username2),
+					resource.TestCheckResourceAttr(resourceName, "user.1.roles.#", "1"),
+					resource.TestCheckTypeSetElemAttr(resourceName, "user.1.roles.*", contributorRole),
```

Comment:
I've not used this - but if you're using a set, and you want to know if your element exists, you can refer to it by it's hashcode here like `user.1.roles.3487564` RT has some examples.

Reply:
Not this particular one. This is a `TypeList` and the ordering is non-deterministic. Hence the use of this check for the set.

Reply:
for list is should absolutely be deterministic - `TypeSet` is not due to hashing

Reply:
It's deterministic, but the ordering is reversed. Not sure which is worse; checking for the 2 values but in reversed ordered, or checking both exists in the set. I'll use the former for now.